src/OpenLoco/src/Map/MapGenerator.cpp

```diff
@@ -138,14 +139,90 @@ namespace OpenLoco::World::MapGenerator
 
     class OriginalTerrainGenerator
     {
-    public:
-        // 0x004624F0
-        void generate([[maybe_unused]] const S5::Options& options, HeightMap& heightMap)
+    private:
+        // 0x00462518
+        void blitHillsByTopographyStyle([[maybe_unused]] const S5::Options& options, HeightMap& heightMap)
         {
             _heightMap = heightMap.data();
-            call(0x004624F0);
+
+            call(0x00462518);
+
             _heightMap = nullptr;
         }
+
+        // 0x00462556
+        void copyHeightMapFromG1(Gfx::G1Element* g1Element, HeightMap& heightMap)
+        {
+            auto* src = g1Element->offset;
+            auto* dst = heightMap.data();
+
+            for (auto y = World::kMapRows; y > 0; y--)
+            {
+                dst += World::kMapColumns;
+
+                for (auto x = World::kMapColumns; x > 0; x--)
+                {
+                    dst--;
+                    *dst = std::max<uint8_t>(*dst, *src);
+                    src++;
+                }
+
+                src += World::kMapPitch;
+            }
+        }
+
+        // 0x00462590
+        void capSeaLevels(HeightMap& heightMap)
+        {
+            auto seaLevel = getGameState().seaLevel;
+            auto* dst = heightMap.data();
+
+            for (auto i = World::kMapPitch * (World::kMapPitch - 1) - 1; i > 0; i--)
+            {
+                if (seaLevel < *(dst))
+                    continue;
+
+                if (seaLevel < *(dst + 1))
+                    continue;
+
+                if (seaLevel < *(dst + World::kMapPitch))
+                    continue;
+
+                if (seaLevel < *(dst + World::kMapPitch + 1))
+                    continue;
+
+                *dst += 1;
+                *(dst + 1) += 1;
+                *(dst + World::kMapPitch) += 1;
+                *(dst + World::kMapPitch + 1) += 1;
+
+                dst++;
+            }
+        }
+
+    public:
+        // 0x004624F0
+        void generate(const S5::Options& options, HeightMap& heightMap)
+        {
+            for (auto i = 0x10000; i > 0; i--)
+            {
+                static const auto baseHeight = options.minLandHeight * 0x1010101;
+                _heightMap[i] = baseHeight;
+            }
```

Comment:
I think its better to move the baseHeight variable out of the loop and use std::fill, it also doesn't need to be static, I looked at the original code as that multiplication looked odd, the original also used memset/rep stosd

Reply:
You're absolutely right; this is a bit cumbersome. I also want to refactor how the `heightMap` data is accessed. Rather than using `.data()` to get to the raw `uint8_t`s, we should really use the point-wise operators that have been defined instead.